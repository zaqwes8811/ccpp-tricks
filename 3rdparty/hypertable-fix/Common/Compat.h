/*
 * Copyright (C) 2007-2012 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/** @file
 * Compatibility Macros for C/C++.
 * This file contains compatibility macros for C/C++. Always include this
 * file before including any other file!
 */

#ifndef HYPERTABLE_COMPAT_H
#define HYPERTABLE_COMPAT_H

/** @addtogroup Common
 *  @{
 */

// The same stuff for C code
#include "compat-c.h"

#include <cstddef> // for std::size_t and std::ptrdiff_t

// C++ specific stuff
#ifndef BOOST_SPIRIT_THREADSAFE
#  define BOOST_SPIRIT_THREADSAFE
#endif

#define BOOST_IOSTREAMS_USE_DEPRECATED
#define BOOST_FILESYSTEM_VERSION 3
#define BOOST_FILESYSTEM_DEPRECATED     1

#include "Sweetener.h"

#define HT_UNUSED(x) static_cast<void>(x)

/** @}*/

#endif // HYPERTABLE_COMPAT_H
