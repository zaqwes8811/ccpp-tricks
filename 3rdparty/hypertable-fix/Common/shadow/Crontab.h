/* -*- c++ -*-
 * Copyright (C) 2007-2013 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Crontab class for periodic events.
/// The Crontab class is used to track the timing of a periodic event.
/// This is used i.e. in the Hypertable.Master to schedule LoadBalancer events.

#ifndef HYPERTABLE_CRONTAB_H
#define HYPERTABLE_CRONTAB_H

#include "String.h"

#include <time.h>

#include <bitset>

namespace Hypertable {

  /// @addtogroup Common
  /// @{

  /// Binary representation of crontab spec.
  struct crontab_entry {
    std::bitset<60> minute;
    std::bitset<24> hour;
    std::bitset<31> dom;
    std::bitset<12> month;
    std::bitset<8>  dow;
  };

  /// Tracks timing of periodic events.
  /// The Crontab class is used to track the timing of a
  /// periodic event.  It is constructed with a string argument
  /// that has the format of first five fields (date/time fields)
  /// of a crontab entry and is used to determine at what time
  /// an event should occcur.
  class Crontab {
  public:

    /// Default constructor.
    Crontab() { };

    /// Constructor initialized with crontab(5) entry.
    /// initializes the Crontab from a string which is formatted
    /// like a regular unix crontab entry:
    /// <pre>
    ///   "1 1 * * * *"
    ///   "1-2/2/4 1 * * *"
    ///   "2-1 1 * * *"
    ///   "1-8/ 1 * * *"
    /// <pre>
    /// @see <code>Common/tests/crontab_test.cc</code> for more examples.
    /// @param spec The crontab string describing the time interval
    /// @throws Error::COMMAND_PARSE_ERROR if the spec cannot be parsed
    Crontab(const String &spec);

    /// Retrieves the timestamp of the next event.
    /// @param now The current time, in seconds since the epoch
    /// @return The absolute time of the next event, in seconds since
    ///      the epoch
    time_t next_event(time_t now);

    /// Get the internal crontab_entry structure.
    /// @note Required for testing
    crontab_entry &entry() { return m_entry; }

  private:

    /// Determines next day on which event will occur.
    /// Using the date passed in through <code>next_tm</code> as the starting
    /// date, modifies it to hold the day in which the next event will occur.
    /// The <code>tm_hour</code>, <code>tm_min</code>, and <code>tm_sec</code>
    /// fields are cleared.
    /// @param next_tm Structure holding starting date and modified to hold the
    ///   next day the event will occur
    /// @param increment Increment <code>next_tm</code> by one day on entry to
    /// function
    void next_matching_day(struct tm *next_tm, bool increment);

    /// Parses a crontab spec into a crontab_entry.
    /// @param spec %Crontab spec
    /// @param _entry Binary crontab specification filled in by function
    void parse_entry(const String &spec, crontab_entry *_entry);

    /// Parses a crontab field and sets corresponding bits in <code>bits</code>.
    /// @tparam N Sizeof  <code>bits</code>
    /// @param spec %Crontab field
    /// @param bits Bit set to populate
    /// @param zero_based Indicates is crontab field is zero or one based.
    template <int N> void parse_range(const String &spec, std::bitset<N> &bits,
            bool zero_based = true);

    crontab_entry m_entry;
  };

  /// Helper function to write crontab_entry to an ostream.
  /// @param os Output stream
  /// @param entry Binary crontab specification structure
  /// @return Output stream (<code>os</code>)
  std::ostream &operator<<(std::ostream &os, const crontab_entry &entry);

  /// Converts binary crontab spec back into string spec.
  /// @note For testing
  /// @tparam N Number of bits in <code>bits</code>
  /// @param bits Bit set representing crontab field
  /// @param spec %Crontab field string equivalent of <code>bits</code>
  template <int N> void reconstruct_spec(const std::bitset<N> &bits,
            String &spec);

  /// @}

} // namespace Hypertable

#endif // HYPERTABLE_CRONTAB_H
