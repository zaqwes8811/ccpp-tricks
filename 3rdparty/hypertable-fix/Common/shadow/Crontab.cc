/*
 * Copyright (C) 2007-2012 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/// @file
/// Crontab class for periodic events.
/// The Crontab class is used to track the timing of a periodic event.
/// This is used i.e. in the Hypertable.Master to schedule LoadBalancer events.

#include <Common/Compat.h>
#include "Crontab.h"

#include <Common/Logger.h>
#include <Common/StringExt.h>

#include <boost/tokenizer.hpp>

#include <cctype>
#include <cstdlib>

using namespace Hypertable;
using namespace boost;
using namespace std;

Crontab::Crontab(const String &spec) {
  parse_entry(spec, &m_entry);
}

time_t Crontab::next_event(time_t now) {
  struct tm next_tm;
  int i;
  int hour_increment = 0;
  int day_increment = 0;

  // align to "next" minute boundary
  now += 59;

  localtime_r(&now, &next_tm);
  next_tm.tm_sec = 0;

 next_day:
  next_matching_day(&next_tm, day_increment);

 next_hour:
  for (i = next_tm.tm_hour + hour_increment; i < 24; i++) {
    if (m_entry.hour[i]) {
      if (i > next_tm.tm_hour) {
        next_tm.tm_hour = i;
        next_tm.tm_min = 0;
      }
      break;
    }
  }
  if (i == 24) {
    day_increment = 1;
    hour_increment = 0;
    goto next_day;
  }

  for (i = next_tm.tm_min; i < 60; i++) {
    if (m_entry.minute[i]) {
      next_tm.tm_min = i;
      return mktime(&next_tm);
    }
  }
  if (i == 60) {
    hour_increment = 1;
    next_tm.tm_min = 0;
    goto next_hour;
  }

  return mktime(&next_tm);
}

void Crontab::next_matching_day(struct tm *next_tm, bool increment) {
  time_t t;
  bool advanced = false;

  if (increment) {
    next_tm->tm_sec = 0;
    next_tm->tm_min = 0;
    next_tm->tm_hour = 0;
    t = mktime(next_tm) + 86400;
    localtime_r(&t, next_tm);
    advanced = true;
  }

  while (true) {
    if (m_entry.month[next_tm->tm_mon] &&
        (m_entry.dom[(next_tm->tm_mday - 1)] || m_entry.dow[next_tm->tm_wday]))
      break;
    t = mktime(next_tm) + 86400;
    localtime_r(&t, next_tm);
    advanced = true;
  }
  if (advanced) {
    next_tm->tm_sec = 0;
    next_tm->tm_min = 0;
    next_tm->tm_hour = 0;
  }
}


void Crontab::parse_entry(const String &spec, crontab_entry *entry) {
  String text = spec;
  String fields[5];
  size_t field_count = 0;
  bool wildcard_dom = false;
  bool wildcard_dow = false;

  {
    char_separator<char> sep(" ");
    tokenizer< char_separator<char> > tokens(text, sep);
    foreach_ht (const string& t, tokens) {
      if (field_count == 5)
        HT_THROW(Error::COMMAND_PARSE_ERROR, "too many fields");
      fields[field_count++] = t;
    }
    if (field_count < 5)
      HT_THROW(Error::COMMAND_PARSE_ERROR, "too few fields");
  }

  // minute
  {
    char_separator<char> sep(",");
    tokenizer< char_separator<char> > tokens(fields[0], sep);
    foreach_ht (const string& t, tokens)
      parse_range<60>(t, entry->minute);
  }

  // hour
  {
    char_separator<char> sep(",");
    tokenizer< char_separator<char> > tokens(fields[1], sep);
    foreach_ht (const string& t, tokens)
      parse_range<24>(t, entry->hour);
  }

  // dom
  {
    if (fields[2] == "*")
      wildcard_dom = true;
    else {
      char_separator<char> sep(",");
      tokenizer< char_separator<char> > tokens(fields[2], sep);
      foreach_ht (const string& t, tokens)
        parse_range<31>(t, entry->dom, false);
    }
  }

  // month
  {
    char_separator<char> sep(",");
    tokenizer< char_separator<char> > tokens(fields[3], sep);
    foreach_ht (const string& t, tokens)
      parse_range<12>(t, entry->month, false);
  }

  // dow
  {
    if (fields[4] == "*")
      wildcard_dow = true;
    else {
      char_separator<char> sep(",");
      tokenizer< char_separator<char> > tokens(fields[4], sep);
      foreach_ht (const string& t, tokens)
        parse_range<8>(t, entry->dow);
      if (entry->dow[7])
        entry->dow[0] = true;
      if (entry->dow[0])
        entry->dow[7] = true;
    }
  }

  if (wildcard_dom) {
    if (wildcard_dow)
      entry->dom.set();
    else
      entry->dom.reset();
  }
  else {
    if (wildcard_dow)
      entry->dow.reset();
  }
  
}

template <int N> 
void Crontab::parse_range(const String &spec, std::bitset<N> &bits,
        bool zero_based) {
  String text = spec;
  String range, step_str;
  size_t count = 0;

  // Split into range & step
  int step = -1;
  {
    char_separator<char> sep("/");
    tokenizer< char_separator<char> > tokens(text, sep);
    foreach_ht (const string& t, tokens) {
      if (count == 0)
        range = t;
      else if (count == 1)
        step_str = t;
      else
        HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");
      count++;
    }

    if (count == 2) {
      if (step_str == "")
        HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");
      for (const char *ptr = step_str.c_str(); *ptr; ptr++) {
        if (!isdigit(*ptr))
          HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");
      }
      step = atoi(step_str.c_str());
    }
  }

  count = 0;
  int start = -1;
  int end = -1;
  bool got_wildcard = false;
  {
    char_separator<char> sep("-");
    tokenizer< char_separator<char> > tokens(range, sep);
    foreach_ht (const string& t, tokens) {
      if (count == 0) {
        if (t == "*") {
          bits.set();
          got_wildcard = true;
        }
        else {
          for (const char *ptr = t.c_str(); *ptr; ptr++) {
            if (!isdigit(*ptr))
              HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");
          }
          start = atoi(t.c_str());
        }
      }
      else if (count == 1) {
        if (got_wildcard)
          HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");
        for (const char *ptr = t.c_str(); *ptr; ptr++) {
          if (!isdigit(*ptr))
            HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");
        }
        end = atoi(t.c_str());
      }
      else
        HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");
      count++;
    }
  }

  if (got_wildcard)
    return;

  int offset = 0;
  if (!zero_based) {
    offset = 1;
    if (start == 0)
      HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");
  }

  if (start == -1)
    HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");

  if (start >= (N+offset) || end >= (N+offset))
    HT_THROW(Error::COMMAND_PARSE_ERROR, "bad specification");


  if (end == -1)
    bits[start-offset] = true;

  if (step == -1)
    step = 1;

  for (int i=(start-offset); i<=(end-offset); i+=step)
    bits[i] = true;
}

std::ostream &Hypertable::operator<<(std::ostream &os,
                                     const Hypertable::crontab_entry &entry) {
  String spec;

  reconstruct_spec<60>(entry.minute, spec);
  os << spec;
  reconstruct_spec<24>(entry.hour, spec);
  os << " " << spec;
  reconstruct_spec<31>(entry.dom, spec);
  os << " " << spec;
  reconstruct_spec<12>(entry.month, spec);
  os << " " << spec;
  reconstruct_spec<8>(entry.dow, spec);
  os << " " << spec;

  return os;
}

template <int N> void Hypertable::reconstruct_spec(const std::bitset<N> &bits,
                                                   String &spec) {
  bool first = true;
  int start = -1;
  int end = -1;
  spec = "";
  for (int i=0; i<N; i++) {
    if (bits[i]) {
      if (start == -1)
        start = i;
      else
        end = i;
    }
    else if (start != -1) {
      if (!first)
        spec += ",";
      first = false;
      if (i == start+1)
        spec += String("") + start;
      else
        spec += String("") + start + "-" + end;
      start = end = -1;
    }
  }
  if (start != -1) {
    if (!first)
      spec += ",";
    if (start == 0 && end == N-1)
      spec += "*";
    else {
      if (end == -1)
        spec += String("") + start;
      else
        spec += String("") + start + "-" + end;
    }
  }
  else if (first)
    spec += "*";
}
