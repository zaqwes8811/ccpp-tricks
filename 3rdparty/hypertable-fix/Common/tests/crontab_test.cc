/** -*- C++ -*-
 * Copyright (C) 2007-2012 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Hypertable. If not, see <http://www.gnu.org/licenses/>
 */

#include "Common/Compat.h"
#include "Common/Crontab.h"
#include "Common/Error.h"

#include <iostream>

#include <boost/algorithm/string.hpp>

using namespace Hypertable;
using namespace std;

namespace {

  const char *specs[] = {
    "1 1 * *",
    "1 1 * * * *",
    "1-2/2/4 1 * * *",
    "2-1 1 * * *",
    "1-8/ 1 * * *",
    "1-8/a 1 * * *",
    "*-8 1 * * *",
    "1-2b 1 * * *",
    "1-2-3 1 * * *",
    "1 * * * *",
    "2-3 * * * *",
    "* 1,2,7 * * *",
    "* 0-22/2 * * *",
    "* 0-21/3 * * *",
    "* * */2 * *",
    "60 * * * *",
    "* 24 * * *",
    "* * 0 * *",
    "* * 32 * *",
    "* * * 0 *",
    "* * * 13 *",
    "* * * 4-6 *",
    "* * 15-30 * *",
    (const char *)0
  };

  void display_next_event(Crontab &crontab, struct tm *tmval) {
    char output_buf[32];
    time_t t, next;
    t = mktime(tmval);
    next = crontab.next_event(t);
    ctime_r(&next, output_buf);
    String output(output_buf);
    boost::trim_if(output, boost::is_any_of("\n\t "));
    cout << output << "\n";
  }

  void display_next_event(Crontab &crontab, time_t t) {
    char output_buf[32];
    time_t next;
    next = crontab.next_event(t);
    ctime_r(&next, output_buf);
    String output(output_buf);
    boost::trim_if(output, boost::is_any_of("\n\t "));
    cout << output << "\n";
  }


}

int main(int argc, char *argv[]) {
  Crontab crontab;

  for (size_t i=0; specs[i]; i++) {
    try {
      cout << Crontab(specs[i]).entry() << "\n";
    }
    catch (Exception &e) {
      cout << e.what() << " (" << specs[i] << ")\n";
    }
  }

  crontab = Crontab("0 0 * * *");
  display_next_event(crontab, (time_t)1352793827);

  crontab = Crontab("0 0 29 2 *");
  cout << crontab.entry() << "\n";

  struct tm tmval;

  memset(&tmval, 0, sizeof(tmval));

  tmval.tm_mday = 2;
  tmval.tm_mon = 8;
  tmval.tm_year = 2012 - 1900;
  display_next_event(crontab, &tmval);

  tmval.tm_mday = 1;
  tmval.tm_mon = 2;
  tmval.tm_year = 2016 - 1900;
  display_next_event(crontab, &tmval);

  tmval.tm_mday = 29;
  tmval.tm_mon = 1;
  tmval.tm_year = 2000 - 1900;
  display_next_event(crontab, &tmval);

  crontab = Crontab("0 0 * * 0");
  cout << crontab.entry() << "\n";

  tmval.tm_mon = 8;
  tmval.tm_year = 2012 - 1900;

  tmval.tm_mday = 3;
  display_next_event(crontab, &tmval);

  tmval.tm_mday = 10;
  display_next_event(crontab, &tmval);

  crontab = Crontab("0 0 * * *");
  cout << crontab.entry() << "\n";

  tmval.tm_mon = 8;
  tmval.tm_year = 2012 - 1900;

  tmval.tm_min = 27;
  tmval.tm_hour = 3;
  tmval.tm_mday = 2;
  display_next_event(crontab, &tmval);

  tmval.tm_min = 9;
  tmval.tm_hour = 12;
  tmval.tm_mday = 3;
  display_next_event(crontab, &tmval);

  tmval.tm_min = 44;
  tmval.tm_hour = 17;
  tmval.tm_mday = 4;
  display_next_event(crontab, &tmval);

  tmval.tm_min = 0;
  tmval.tm_hour = 12;
  tmval.tm_mday = 5;
  display_next_event(crontab, &tmval);

  crontab = Crontab("30 12 3,5 * *");
  cout << crontab.entry() << "\n";

  tmval.tm_mon = 8;

  tmval.tm_min = 0;
  tmval.tm_hour = 15;
  tmval.tm_mday = 3;
  display_next_event(crontab, &tmval);

  tmval.tm_min = 0;
  tmval.tm_hour = 12;
  tmval.tm_mday = 3;
  display_next_event(crontab, &tmval);

  tmval.tm_min = 30;
  tmval.tm_hour = 12;
  tmval.tm_mday = 3;
  display_next_event(crontab, &tmval);

  tmval.tm_min = 31;
  tmval.tm_hour = 12;
  tmval.tm_mday = 3;
  display_next_event(crontab, &tmval);

  return 0;
}
