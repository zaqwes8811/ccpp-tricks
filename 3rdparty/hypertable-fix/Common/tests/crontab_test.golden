too few fields (1 1 * *)
too many fields (1 1 * * * *)
bad specification (1-2/2/4 1 * * *)
* 1 * * *
1-8 1 * * *
bad specification (1-8/a 1 * * *)
bad specification (*-8 1 * * *)
bad specification (1-2b 1 * * *)
bad specification (1-2-3 1 * * *)
1 * * * *
2-3 * * * *
* 1-2,7 * * *
* 0,2,4,6,8,10,12,14,16,18,20,22 * * *
* 0,3,6,9,12,15,18,21 * * *
* * * * *
bad specification (60 * * * *)
bad specification (* 24 * * *)
bad specification (* * 0 * *)
bad specification (* * 32 * *)
bad specification (* * * 0 *)
bad specification (* * * 13 *)
* * * 3-5 *
* * 14-29 * *
Wed Nov 14 00:00:00 2012
0 0 28 1 *
Mon Feb 29 00:00:00 2016
Sat Feb 29 00:00:00 2020
Tue Feb 29 00:00:00 2000
0 0 * * 0,7
Sun Sep  9 00:00:00 2012
Sun Sep 16 00:00:00 2012
0 0 * * *
Mon Sep  3 00:00:00 2012
Tue Sep  4 00:00:00 2012
Wed Sep  5 00:00:00 2012
Thu Sep  6 00:00:00 2012
30 12 2,4 * *
Wed Sep  5 12:30:00 2012
Mon Sep  3 12:30:00 2012
Mon Sep  3 12:30:00 2012
Wed Sep  5 12:30:00 2012
